### PR TITLE
use separate channel to signal session close

### PR DESCRIPTION
**Pull request by joesis**
Should fix https://github.com/getlantern/flashlight/issues/363

In the current code, `buf.ack`(aka `session.out`) may be closed before sending to it https://github.com/getlantern/lampshade/blob/master/receive_buffer.go#L148. This introduces a `closeCh` so the data channel never closes.

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -252,12 +255,10 @@ func (s *session) sendLoop() {
...
 		select {
-		case frame, open := <-s.out:
-			stillOpen := true
-			if frame != nil {
-				stillOpen = s.send(frame)
-			}
-			if !open || !stillOpen {
+		case <-s.closeCh:
+			return
+		case frame := <-s.out:
```
Comment (oxtoacart) [inline session.go]: The problem with using a separate channel is that in theory, we could select from closeCh and return even though there's buffered data in s.out remaining to be sent. I'll try a different approach for fixing.

**Comment by oxtoacart** [comment]
See #12 for an alternative approach that allows us to close the `out` channel as before.

File: session.go (hunk 1 above)
Comment (joesis) [reply session.go]: How about draining the channel until it becomes empty (both here and in `coalesceAdditionalFrames`)? It feels to me that having a separate close channel made the code a bit more readable.

Comment (oxtoacart) [reply session.go]: Yes, draining should work.

**Comment by oxtoacart** [comment]
Okay, this LGTM!

**Comment by joesis** [comment]
Thanks @oxtoacart

**Pull request merged**